### DISCUSSION TITLE
New keyword: exists

### DISCUSSION THREAD

**Discussion by Johnhersh** [dotnet/csharplang#6740] [Language Ideas]
I would like something like:
`if (variable exists)...`
Basically the same as doing:
`if (variable is not null)...`
The problem we're having now is we've switched from using '!= null' to 'is not null' because it's easier to read. But this also introduces a double-negative. So team members choose to go for 'is { }' which we all agree is less readable, but has the advantage of being a positive instead of a double negative. It would be nice to replace that with just 'exists'

**Comment by 333fred** [comment]
`is not null` is a single negative, not a double negative.

**Reply by Johnhersh** [reply]
null is saying "object doesn't exist" so it has an implied negativity to it. In plain English, what we believe on my team, is that "is not null" means "object doesn't not exist".

From wikipedia on null pointer:
In computing, a null pointer or null reference is a value saved for indicating that the pointer or reference does not refer to a valid object.

So "is not null" means "doesn't not refer to a valid object"

**Reply by HaloFour** [reply]
C# doesn't follow English grammar rules, so there's no reason to avoid what you'd consider to be a double negative.

**Reply by Johnhersh** [reply]
I understand that, I also believe it would be nicer to have the option to work that way. It seems to me it's along the same lines as them adding "is not null" when "!= null" was already there, isn't it?

**Reply by jnm2** [reply]
It's not really along the same lines, since the goal was not to provide an identical alternative to `!= null`. The goal was to add logical operations to patterns (`and`, `or`, and `not`) and `not null` fell out without being a necessary goal. For instance: `x is not Foo { Bar: 42 }`.

(Also, `is not null` and `!= null` do not mean exactly the same thing. The latter calls overloaded operators if they exist, the former is a reference equality check.)

**Reply by HaloFour** [reply]
The team didn't specifically add `is not null`, they added `not` as a modifier to negate patterns, and it can be used in many other scenarios.  It happens to be used in this context because `null` is a pattern` and the `is` operator is how you compare a value to a pattern.  This is similar to `{ }` which is really a degenerate form of property patterns where the type check is omitted and there are no property patterns.

**Reply by Johnhersh** [reply]
Fair enough. Still though now that we have stuff like "is" and "not" and "is not null", what's the harm in having "exists" as an alternative to "is { }" ?

The feature is already there. You can already say "is { }", so how is allowing "is { }" but not "exists" beneficial to us if "exists" will increase readability slightly?

**Reply by jnm2** [reply]
"What's the harm" is not a motivator to prioritize something above all the other things the C# team could be working on.

**Reply by HaloFour** [reply]
The question is whether or not it's worth the effort.

There are a number of ways to make this comparison, some of which have been the idiomatic form for 20 years in C# and many more in C-like languages.  Whereas most of the other new ways to compare fell out of larger features, this proposal only has one narrow purpose so the gain is a lot less.  I don't agree with the "readability" concerns, I find `x is not null` and `x != null` both extremely easy to read.  So it comes down to a stylistic preference for C# to read more like English, but C# isn't attempting to follow English grammar rules.

**Reply by Johnhersh** [reply]
Fair enough again, I'm not on the team so it's easy for me to make a suggestion. I cannot speak to prioritization.

Still seems like a valid suggestion to me, if not a very urgent one. I'm sure I'm not the only one who would like it. It's also something used in SQL so it's not a brand new idea. And I could see it being used in LINQ to translate into the sql EXISTS

**Reply by RikkiGibson** [reply]
I am personally not thrilled that "null-check and create an implicitly typed variable" is currently best expressed as `expr is { } x`. But I don't really see a way of improving it.

**Reply by Johnhersh** [reply]
Wouldn’t “exists” as syntactic sugar for “is { }” solving that?

**Reply by CyrusNajmabadi** [reply]
I prefer `is not null` since it's exactly how everyone i know talks when just describing values in execution flow.  So the c# code just matches that.  in other words, i don't hear people say "if x exists, then do blah" they say "if x isn't null, do blah".

**Reply by En3Tho** [reply]
I don't think it would. 
```if (expr exists x)``` feels like super weird.
It doesn't combine with anything to be honest. And there is zero advantage to ```if (Exists(expr)) ...```

**Reply by DavidArno** [reply]
> So "is not null" means "doesn't not refer to a valid object"

And a basic rule of English is that the double negative cancels out, so "is not null" means "refers to a valid object", which is how I at least read "is not null". And I suspect the number of teams choosing to see it as a double negative and thus a problem is vanishingly rare.

But the killer reason for not wanting to see this change occur is the simple fact that "exists" does not mean "refers to a valid object". So when seeing `if (x exists)`, I'm trying to think of a situation where the variable `x` doesn't exist, ie hasn't been defined. So I could see it applying to a language wrapper over reflection, eg `if (exists(Foo.Bar))` to test whether `Foo` has a member `Bar`, but using it as a test for not null makes no sense to me at all.

**Reply by CPosingies** [reply]
The thing is, ever since C# left the "Never mind Whitespace" C-community (remember (if(x|z>>2&&x+a))return0; times), it has in fact become more readable, clearer, easier to learn. Still, there is an overwhelmingly acclaimed and back-from-the-K&R-times-bullet-proven basic framework regarding C-derived languages (well, except Java) that is when it's going hardcore, it should be written hardcore. "We choose to to write this code in this decade and do the other things not because they are easy, but because they are hard!"

Why are there !, &, |, &&, ||, <<, ^ in C#, still? Because they are fundamentally precice and clear. Because you can, even without "using System;", write a whole Operating System in C#.

Exhausting... Now, back to your suggestion.

What would "if (foo exists)" mean?

int foo;
if (foo exists) return true;

int? foo;
if (foo exists) return true;

object o = null;
if (o exists) return true;

Func<double, double> f;
if (f exists) return true;

Back to the former topic. I in fact am pretty much into literal programming. It's about the spirit of naming methods and properties so that it pleases the eye of a code's reader. And it is about more code, less comments. Storage space is not a fence anymore. I do love Extension Methods when they make things clearer, even if they only add syntactical sugar -- but LINQ's added a lot of clarity already.

Still, now to your employees. In C-ish languages it's "!", not "NOT" inverting a Boolean expression. It is "a++", not "a = a + 1". If you insist on your employees coding in C#, you'd better hire folks that are able to get !0 = 1, !1 = 0, !!1 = 1. Make it so!

Syntactical sugar should be applied if and only if it'll add to basic comprehension. To me, an "exists" thingy would be confusing. Scroll up for reasons.

**Reply by Johnhersh** [reply]
I see everyone’s points, and while I obviously don’t agree I respect them. However I’m still not getting the resistance in light of the fact that C# already has “if (x is { })” which even in this thread people admitted isn’t the best. 

To say syntactic sugar should only be applied if it’ll make things clearer, then rejecting this, means that somehow “is { }” is clearer than “exists”.

**Reply by HaloFour** [reply]
@Johnhersh 

> To say syntactic sugar should only be applied if it’ll make things clearer, then rejecting this, means that somehow “is { }” is clearer than “exists”.

The team didn't add `is { }`, they added `is Type { Property: pattern }` where the `Type` pattern can be elided when known, and supporting 0.._n_ property subpatterns.  The form `is { }` fell out as an accident, and wasn't considered offensive enough to intentionally forbid.

I think that the team would agree that `is { }` is not very clear, which is part of the motivation behind supporting `not` as a negation pattern when combined with the `null` pattern.

But I think it's important that both property patterns and negation patterns can be used in multiple ways to solve multiple problems.

**Reply by Johnhersh** [reply]
>But the killer reason for not wanting to see this change occur is the simple fact that "exists" does not mean "refers to a valid object". So when seeing if (x exists), I'm trying to think of a situation where the variable x doesn't exist, ie hasn't been defined. So I could see it applying to a language wrapper over reflection, eg if (exists(Foo.Bar)) to test whether Foo has a member Bar, but using it as a test for not null makes no sense to me at all.

This is the best counterargument I've heard. It makes total sense. To me it just means I've picked the wrong keyword here. In other environments you can do something like `if (foo)` without any operands and that's also a null check, and I would much prefer that over `if (foo is { })`

>I think that the team would agree that is { } is not very clear, which is part of the motivation behind supporting not as a negation pattern when combined with the null pattern.
>
>But I think it's important that both property patterns and negation patterns can be used in multiple ways to solve multiple problems.

Then the issue here is that 'exists' is too narrow of a solution? You'd like to see something that's also usable in other cases?

**Reply by CyrusNajmabadi** [reply]
> However I’m still not getting the resistance in light of the fact that C# already has “if (x is { })” which even in this thread people admitted isn’t the best.

In a very real sense it's not enough for features just to be okay. They have to truly solve a real issue that we consider significant.  Making an existing construct only ever so slightly better does not rise to that level. 

Note: conversations here can be confusing as they cover two concepts. First, the idea itself (whether it is sound, does the design actually work), and then whether or not it makes sense to actually do the work and add to the language. 

Conceptually, there is no problem with a shorthand for `not null`. So it's passes that test. But as to if this is valuable enough for language, it falls very short.

**Reply by Johnhersh** [reply]
>Conceptually, there is no problem with a shorthand for not null. So it's passes that test. But as to if this is valuable enough for language, it falls very short.

Ok I see, that makes sense. Thanks for explaining

**Comment by Xyncgas** [comment]
I like the new keyword, I think .NET team just too busy implementing it
but if you stop using c# you can use F#
```F#
let k = if exists myVariable then 0 else 1
```
or
```F#
let k =
    try
        match myVariable with
        | Some value -> 0
        | None -> 1
    with | _ -> 1
```

**Reply by Johnhersh** [reply]
Oh that's exactly what I was looking for! I had no idea F# had that
